Get additional info on deployment failures

krithika369 opened the pull request:
This PR implements the API and UI changes for storing additional info on deployment failures in the DB and displaying them as a part of the `Activity Log` panel in the Turing UI.

### API

When a Knative service fails to deploy within the configured timeout, we attempt to get one of the following logs from the Kubernetes resources, before purging them:
* `user-container` termination message - When a container fails to become ready and the pod repeatedly restarts, the termination message holds the relevant logs ([ref](https://github.com/knative/serving/blob/master/pkg/reconciler/revision/resources/deploy.go#L151))
* Knative service status condition - In cases where the pod was never created (such as bad image name for the user-container), the Knative service's statuses may contain some useful info. So, we retrieve all the status transitions from the service.

These changes are implemented in `api/turing/cluster/controller.go`.

### UI

Since container termination logs can be long (up to 2048 characters), the Activity Log panel has been modified to allow for collapsing the logs. By default, only the first 3 lines are visible and the rest can be seen by expanding the log. To achieve this, the message is now being wrapped into an `ExpandableContainer`.
* `ui/src/components/expandable_container/ExpandableContainer.js` - Add properties `isScrollable` (decides whether the content is scrollable and a scrollbar should be added) and `toggleKind` (either "button" or "text", which will use `EuiButton` and `EuiLink`, respectively).

**isScrollable=true, toggleKind="button" (default)**
![Screenshot 2020-12-07 at 3 56 41 PM](https://user-images.githubusercontent.com/23465343/101324732-a988b480-38a5-11eb-94fb-421d4cb87dd3.png)

**isScrollable=false toggleKind="text" (event item collapsed)**
![Screenshot 2020-12-07 at 3 57 23 PM](https://user-images.githubusercontent.com/23465343/101324748-af7e9580-38a5-11eb-8a67-fa87212830b5.png)

**isScrollable=false toggleKind="text" (event item expanded)**
![Screenshot 2020-12-07 at 3 57 38 PM](https://user-images.githubusercontent.com/23465343/101324784-b9a09400-38a5-11eb-941d-889c968fd187.png)

* `ui/src/router/activity_log/events_list/EventsList.js` - Wrap event message into an `ExpandableContainer`
* Styling changes so that the Version History and Activity Log panels are of equal width.

krithika369 commented on ui/src/router/activity_log/events_list/EventsList.js:
```diff
@@ -30,7 +31,9 @@ export const EventsList = ({ events, status }) => {
           type="update"
           timestamp={<DateFromNow date={event.created_at} size={"s"} />}>
-          {event.message}
+          <ExpandableContainer maxHeight={43} buttonSize="s">
```
Height=43 shows 3 lines of text nicely.

krithika369 commented on ui/src/router/components/configuration/components/section.scss:
```diff
@@ -4,7 +4,7 @@
 // is expected to be set to a fixed value.
 // Ref: https://css-tricks.com/flexbox-truncated-text/
-.euiFlexItem.euiFlexItem--childFlexPanel {
+.euiFlexItem--childFlexPanel {
```
`.euiFlexItem.euiFlexItem--childFlexPanel` is unnecessarily restrictive, should be enough to look at "euiFlexItem--childFlexPanel".

krithika369 commented on ui/src/router/history/HistoryView.js:
```diff
@@ -6,11 +6,11 @@ import { ListRouterVersionsView } from "../versions/list/ListRouterVersionsView"
   return (
     <EuiFlexGroup direction="row">
-      <EuiFlexItem grow={5}>
+      <EuiFlexItem grow={1} className="euiFlexItem--mediumPanel">
```
Without the minimum width property applied by  `euiFlexItem--mediumPanel`, Activity Log panel grows to occupy most of the space.

davidheryanto commented on api/turing/cluster/controller.go:
```diff
@@ -488,7 +489,23 @@ func (c *controller) waitKnativeServiceReady(
+				} else {
+					failedCondition := svc.Status.GetCondition(knativeapis.ConditionType("ConfigurationsReady"))
+					if failedCondition != nil && failedCondition.IsFalse() {
+						terminationMessage = failedCondition.GetMessage()
```
Maybe we can use this const for the "ConfigurationsReady" condition type?
https://github.com/knative/serving/blob/7398ed7b7e45bb751e6aa475a8d2f2ab15398e44/pkg/apis/serving/v1/service_types.go#L101

I think the variable name `failedCondition` can be misinterpreted, one would think `failedCondition.IsFalse()` means condition is successful hence the configuration is Ready, when it is actually not.

Can we maybe use `configurationReadyCondition` or something similar as the variable name?

davidheryanto commented on api/turing/cluster/controller.go:
```diff
@@ -488,7 +489,23 @@ func (c *controller) waitKnativeServiceReady(
+					if failedCondition != nil && failedCondition.IsFalse() {
+						terminationMessage = failedCondition.GetMessage()
+					} else {
+						terminationMessage = "cause unknown"
```
In this case, in addition to saying the cause is unknown, can we maybe also provide a bit more context, maybe just listing all the Condition statuses for the Knative service object?

krithika369 replied:
Thanks, David. I changed this so that we dump all the status conditions from the service if the pod termination message is not available.

krithika369 commented on ui/src/components/expandable_container/ExpandableDescriptionList.scss:
```diff
@@ -2,7 +2,7 @@
 .expandableContainer__childWrapper {
   height: 0;
-  overflow-y: auto;
+  overflow: hidden;
```
Without this, we see unnecessary scroll:

![Screenshot 2020-12-03 at 12 32 17 PM](https://user-images.githubusercontent.com/23465343/100964744-39f08d80-3564-11eb-9277-fe60fc2cb993.png)

![Screenshot 2020-12-03 at 12 30 53 PM](https://user-images.githubusercontent.com/23465343/100964754-3f4dd800-3564-11eb-90fa-8799c84c36be.png)

krithika369 replied:
Yeah the second screenshot is from the Docker Env Variables. If you want to see the scroll there (I thought it was weird to have that and the Expand option), I will apply this style change only for the activity log's expandable container.

krithika369 replied:
Now the scroll behavior is controlled by the presence of the class `scrollableContainer`.

krithika369 commented on ui/src/router/activity_log/events_list/EventsList.scss:
```diff
@@ -0,0 +1,3 @@
+.activityLogEvent {
```
This is so that the "\n" in the activity log event is rendered as a newline (every Type.. is a new line).
![Screenshot 2020-12-03 at 4 18 14 PM](https://user-images.githubusercontent.com/23465343/100982938-9fa04200-3583-11eb-9d81-07bd5a3681b9.png)

romanwozniak reviewed (commented):
Cool, that you re-used existing component, I've added some comments regarding this. 

But ideally I'd prefer to have this implemented as "Show More" link (similar to this component https://www.npmjs.com/package/react-show-more-text). This is more obvious UX compared to an arrow-styled button next to text.

krithika369 commented on ui/src/components/expandable_container/ExpandableContainer.js:
```diff
@@ -1,12 +1,23 @@
 import useDimension from "../../hooks/useDimension";
 
-export const ExpandableContainer = ({ maxHeight = 300, children }) => {
+export const ExpandableContainer = ({
```
Docker env var container will use button toggle + scrollbar. Activity log will use text toggle + no scroll.

krithika369 commented:
@romanwozniak another look for the UI changes? I think I've addressed all your comments. Updated the PR description with new screenshots as well. Thanks!

romanwozniak commented on ui/src/components/expandable_container/ExpandableContainer.js:
```diff
@@ -16,25 +27,38 @@ export const ExpandableContainer = ({ maxHeight = 300, children }) => {
-        }}>
-        <div ref={contentRef}>{children}</div>
-      </div>
+      <EuiFlexGroup direction="row" gutterSize="xs" alignItems="left">
```
Do you still need to wrap this into a flex group?

krithika369 replied:
Removed.

romanwozniak reviewed (approved):
LGTM
